File: 백준/Gold/1074. Z/Z.cc
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

int search(int N,int r,int c)
{
	int k; // 어떤면에 있는지(0~3)
	int minusR = 0, minusC = 0;

	int powResult = pow(2, N - 1);
	//0,1
	if (powResult > r)
	{
		if (powResult > c)
			k = 0;
		else
		{
			k = 1;
			minusC = powResult;
		}
	}
	else //2,3
	{
		if (powResult > c)
			k = 2;
		else
		{
			k = 3;
			minusC = powResult;
		}
		minusR = powResult;
	}

	int tmp = pow(powResult, 2) * k;

	if (N == 1)
		return k;

	return tmp + search(N - 1, r - minusR, c - minusC);
}

int main()
{
	ios::sync_with_stdio(false);
	cin.tie();
	cout.tie();

	int N, r, c;
	cin >> N >> r >> c;

	//2^n * k
	//2^n-1 * k
	//2^n-2 * k
	//다 더하면 답 나올듯
	cout << search(N, r, c) << endl;

	return 0;
}

File: 백준/Gold/1074. Z/Z.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CordingStudy_CS.Week10_CS
{
    internal class Program
    {
        public static void Main(string[] args)
        {
            string[] input = Console.ReadLine().Split();
            int N = int.Parse(input[0]);
            int r = int.Parse(input[1]);
            int c = int.Parse(input[2]);

            Console.WriteLine(Search(N,r,c));
        }


        public static int Search(int N,int r,int c)
        {
            int k;
            int minusR = 0, minusC = 0;

            double powResult = Math.Pow(2, N - 1);

            //0,1면
            if(powResult > r)
            {
                if (powResult > c)
                    k = 0;
                else
                {
                    k = 1;
                    minusC = (int)powResult;
                }
            } //2,3면
            else
            {
                if (powResult > c)
                    k = 2;
                else
                {
                    k = 3;
                    minusC = (int)powResult;
                }
                minusR = (int)powResult;
            }

            int tmp = (int)Math.Pow(powResult, 2) * k;
            if (N == 1)
                return k;
            return tmp + Search(N - 1, r - minusR, c - minusC);
        }
    }
}
